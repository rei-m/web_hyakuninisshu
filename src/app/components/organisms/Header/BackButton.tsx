'use client';

import IconButton from '@mui/material/IconButton';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';
import { useRouter } from 'next/navigation';

export type BackButtonProps = {
  backUrl: string;
};

export const BackButton = ({ backUrl }: BackButtonProps) => {
  const router = useRouter();
  return (
    <IconButton
      onClick={() => {
        router.replace(backUrl, { scroll: true });
      }}
      size="large"
      edge="start"
      color="inherit"
      aria-label="back"
      sx={{ mr: 2 }}
    >
      <ArrowBackIcon sx={{ color: 'common.white', fontSize: '2.4rem' }} />
    </IconButton>
  );
};
