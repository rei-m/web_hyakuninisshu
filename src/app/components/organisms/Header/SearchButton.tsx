'use client';

import IconButton from '@mui/material/IconButton';
import SearchIcon from '@mui/icons-material/Search';
import { useAppDispatch } from '@/lib/hooks';
import { openFilter } from '@/lib/features/material/materialSlice';

export const SearchButton = () => {
  const dispatch = useAppDispatch();

  return (
    <IconButton
      size="large"
      edge="start"
      color="inherit"
      aria-label="menu"
      onClick={() => {
        dispatch(openFilter());
      }}
    >
      <SearchIcon sx={{ color: 'common.white', fontSize: '2.4rem' }} />
    </IconButton>
  );
};
