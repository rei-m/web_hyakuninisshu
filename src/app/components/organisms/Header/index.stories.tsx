import { Meta, StoryObj } from '@storybook/react';
import { Header } from './index';

export default {
  title: 'organisms/Header',
  component: Header,
  parameters: {
    nextjs: {
      appDirectory: true,
    },
  },
} as Meta<typeof Header>;

type Story = StoryObj<typeof Header>;

export const Default: Story = {
  args: {
    title: 'タイトル',
  },
};

export const WithBack: Story = {
  args: {
    ...Default.args,
    backUrl: '/',
  },
};

export const WithSearch: Story = {
  args: {
    ...Default.args,
    isDisplaySearch: true,
  },
};
