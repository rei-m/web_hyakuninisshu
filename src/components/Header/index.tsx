import * as React from 'react';
import styled from '@src/styles/styled-components';
import { withRipple } from '@src/enhancers/withRipple';

export interface Props {
  title: string;
  onClickBack?: () => void;
}

const Container = styled.header<{ canBack: boolean }>`
  box-shadow: ${({ theme }) => theme.elevationShadow1x};
  left: 0;
  right: 0;
  top: 0;
  z-index: 1;
  position: fixed;
  background-color: ${({ theme }) => theme.colorPrimary};
  padding-left: ${({ canBack, theme }) => (canBack ? '0' : theme.spacing2x)};
  display: flex;
`;

const Title = styled.h1`
  font-size: 1.8rem;
  color: #fff;
  line-height: ${({ theme }) => theme.headerHeight};
  margin: 0;
  font-weight: normal;
  text-align: left;
  @media screen and (min-width: ${({ theme }) => theme.minWidthWide}) {
    line-height: ${({ theme }) => theme.headerHeightWide};
  }
`;

const Icon = withRipple(styled.i`
  line-height: ${({ theme }) => theme.headerHeight};
  width: ${({ theme }) => theme.headerHeight};
  color: #fff;
  text-align: center;
  cursor: pointer;
  @media screen and (min-width: ${({ theme }) => theme.minWidthWide}) {
    line-height: ${({ theme }) => theme.headerHeightWide};
    width: ${({ theme }) => theme.headerHeightWide};
  }
`);

const Header: React.FC<Props> = ({ title, onClickBack }) => (
  <Container canBack={!!onClickBack}>
    {!!onClickBack && (
      <Icon className="material-icons" onClick={onClickBack} data-test="back">
        arrow_back
      </Icon>
    )}
    <Title>{title}</Title>
  </Container>
);

export default Header;
