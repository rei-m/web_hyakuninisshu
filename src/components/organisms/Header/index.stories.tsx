import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Header from './index';

storiesOf('organisms/Header', module)
  .add('default', () => <Header title={`へっだー`} />)
  .add('with back button', () => <Header title={`へっだー`} onClickBack={action('onClickBack')} />)
  .add('with search button', () => <Header title={`へっだー`} onClickSearch={action('onClickSearch')} />);
