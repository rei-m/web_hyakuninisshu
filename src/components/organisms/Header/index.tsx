import * as React from 'react';
import ArrowBack from '@material-ui/icons/ArrowBack';
import Search from '@material-ui/icons/Search';
import styled from '@src/styles/styled-components';
import { withRipple } from '@src/enhancers/withRipple';
import Heading from '@src/components/atoms/Heading';
import HeaderItem from '@src/components/molecules/HeaderItem';

export interface Props {
  title: string;
  className?: string;
  onClickBack?: () => void;
  onClickSearch?: () => void;
}

const Container = styled.header<{ canBack: boolean }>`
  box-shadow: ${({ theme }) => theme.elevationShadowHeader};
  background-color: ${({ theme }) => theme.colorPrimary};
  padding-left: ${({ canBack, theme }) => (canBack ? '0' : theme.spacing2x)};
  display: flex;
`;

const StyledHeaderItem = withRipple(styled(HeaderItem)`
  width: ${({ theme }) => theme.headerHeight};
  height: ${({ theme }) => theme.headerHeight};
  @media screen and (min-width: ${({ theme }) => theme.minWidthWide}) {
    width: ${({ theme }) => theme.headerHeightWide};
    height: ${({ theme }) => theme.headerHeightWide};
  }
`);

const Title = styled(Heading)`
  flex-grow: 1;
  color: #fff;
  margin: 0;
  text-align: left;
  line-height: ${({ theme }) => theme.headerHeight};
  @media screen and (min-width: ${({ theme }) => theme.minWidthWide}) {
    line-height: ${({ theme }) => theme.headerHeightWide};
  }
`;

const renderArrowBack = () => <ArrowBack style={{ fontSize: '2.4rem' }} />;
const renderSearch = () => <Search style={{ fontSize: '2.4rem' }} />;

const Header = ({ title, className, onClickBack, onClickSearch }: Props) => (
  <Container canBack={!!onClickBack} className={className}>
    {!!onClickBack && <StyledHeaderItem renderIcon={renderArrowBack} onClick={onClickBack} data-test="back" />}
    <Title level={1} visualLevel={2}>
      {title}
    </Title>
    {!!onClickSearch && <StyledHeaderItem renderIcon={renderSearch} onClick={onClickSearch} data-test="search" />}
  </Container>
);

export default Header;
