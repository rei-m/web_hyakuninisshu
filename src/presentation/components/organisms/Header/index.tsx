import React from 'react';
import clsx from 'clsx';
import makeStyles from '@material-ui/core/styles/makeStyles';
import { SvgIconProps } from '@material-ui/core/SvgIcon';
import ArrowBack from '@material-ui/icons/ArrowBack';
import Search from '@material-ui/icons/Search';
import HeaderItem from '@src/presentation/components/molecules/HeaderItem';
import Heading from '@src/presentation/components/atoms/Heading';
import { withRipple } from '@src/presentation/enhancers/withRipple';
import { ThemeInterface } from '@src/presentation/styles/theme';

export type Props = {
  title: string;
  className?: string;
  onClickBack?: () => void;
  onClickSearch?: () => void;
};

const useStyles = makeStyles<ThemeInterface, { canBack: boolean }>(theme => ({
  root: {
    boxShadow: theme.elevationShadowHeader,
    backgroundColor: theme.palette.primary.main,
    paddingLeft: ({ canBack }) => (canBack ? 0 : theme.spacing(2)),
    display: 'flex',
  },
  title: {
    flexGrow: 1,
    color: '#fff',
    margin: 0,
    textAlign: 'left',
    lineHeight: theme.headerHeight,
    [`@media screen and (min-width:${theme.minWidthWide})`]: {
      lineHeight: theme.headerHeightWide,
    },
  },
  headerItem: {
    width: theme.headerHeight,
    height: theme.headerHeight,
    [`@media screen and (min-width:${theme.minWidthWide})`]: {
      width: theme.headerHeightWide,
      height: theme.headerHeightWide,
    },
  },
}));

type StyledHeaderItemProps = {
  id: string;
  renderIcon: () => React.ReactElement<SvgIconProps>;
  classes: { headerItem: string };
  onClick: () => void;
  className?: string;
};

const StyledHeaderItem = withRipple(({ id, renderIcon, classes, onClick, className = '' }: StyledHeaderItemProps) => {
  return (
    <HeaderItem
      renderIcon={renderIcon}
      onClick={onClick}
      data-test={id}
      className={clsx(classes.headerItem, className)}
    />
  );
});

const renderArrowBack = () => <ArrowBack style={{ fontSize: '2.4rem' }} />;
const renderSearch = () => <Search style={{ fontSize: '2.4rem' }} />;

const Header = ({ title, className = '', onClickBack, onClickSearch }: Props) => {
  const classes = useStyles({ canBack: !!onClickBack });
  return (
    <header className={clsx(classes.root, className)}>
      {!!onClickBack && (
        <StyledHeaderItem
          id="back"
          renderIcon={renderArrowBack}
          onClick={onClickBack}
          classes={{ headerItem: classes.headerItem }}
        />
      )}
      <Heading level={1} visualLevel={2} className={classes.title}>
        {title}
      </Heading>
      {!!onClickSearch && (
        <StyledHeaderItem
          id="search"
          renderIcon={renderSearch}
          onClick={onClickSearch}
          classes={{ headerItem: classes.headerItem }}
        />
      )}
    </header>
  );
};

export default Header;
