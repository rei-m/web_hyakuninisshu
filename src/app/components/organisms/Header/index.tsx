import AppBar from '@mui/material/AppBar';
import Toolbar from '@mui/material/Toolbar';
import Heading from '@/app/components/atoms/Heading';
import SearchButton from '@/app/containers/organisms/Header/SearchButton';
import BackButton from './BackButton';

export type HeaderProps = {
  title: string;
  backUrl?: string;
  isDisplaySearch?: boolean;
};

const Header = (props: HeaderProps) => (
  <AppBar elevation={1} sx={{ zIndex: 3 }}>
    <Toolbar>
      {!!props.backUrl && <BackButton backUrl={props.backUrl} />}
      <Heading level={1} sx={{ flexGrow: 1, color: 'common.white', fontWeight: 700 }}>
        {props.title}
      </Heading>
      {props.isDisplaySearch && <SearchButton />}
    </Toolbar>
  </AppBar>
);

export default Header;
