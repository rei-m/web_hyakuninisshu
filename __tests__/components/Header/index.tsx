import * as ReactTestRenderer from 'react-test-renderer';
import * as React from 'react';
import { shallow } from 'enzyme';
import { sel } from '@test/helpers';
import Header, { Props } from '@src/components/Header';

describe('<Header />', () => {
  let baseProps: Props;

  beforeEach(() => {
    baseProps = {
      onClickBack: jest.fn(),
      title: 'This is title',
    };
  });

  it('should render component when can not back', () => {
    const renderer = ReactTestRenderer.create(<Header {...baseProps} />);
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should render component when can back', () => {
    const props = { ...baseProps, canBack: true };
    const renderer = ReactTestRenderer.create(<Header {...props} />);
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should fire onClickBack when back clicked', () => {
    const props = { ...baseProps, canBack: true };
    const wrapper = shallow(<Header {...props} />);
    wrapper.find(sel('back')).simulate('click');
    expect(baseProps.onClickBack).toHaveBeenCalled();
  });
});
