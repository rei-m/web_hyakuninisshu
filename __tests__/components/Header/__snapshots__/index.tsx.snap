// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render component when can back 1`] = `
<header
  className="sc-bwzfXH geYBAS"
>
  <div
    className="sc-bdVaJa flBAph"
    onMouseDown={[Function]}
    onMouseUp={[Function]}
  >
    <i
      className="material-icons sc-bxivhb jrBfKJ"
      color="#fff"
      data-test="back"
      onClick={[MockFunction]}
    >
      arrow_back
    </i>
  </div>
  <h1
    className="sc-htpNat fxoVCB"
  >
    This is title
  </h1>
</header>
`;

exports[`<Header /> should render component when can not back 1`] = `
<header
  className="sc-bwzfXH geYBAS"
>
  <div
    className="sc-bdVaJa flBAph"
    onMouseDown={[Function]}
    onMouseUp={[Function]}
  >
    <i
      className="material-icons sc-bxivhb jrBfKJ"
      color="#fff"
      data-test="back"
      onClick={[MockFunction]}
    >
      arrow_back
    </i>
  </div>
  <h1
    className="sc-htpNat fxoVCB"
  >
    This is title
  </h1>
</header>
`;
